File: SessionRuntime/src/session_registry.zig
const std = @import("std");
const StateDirectory = @import("state_directory.zig").StateDirectory;
const identities = @import("service_identity.zig");
const client = @import("service_client.zig");
const launch = @import("service_launch.zig");

pub const maximum_sessions: usize = 1024;
pub const maximum_name: usize = 128;
/// Bounded per-session probe. A stopped or wedged session must not stall the
/// whole listing, and the registry never infers liveness from the directory.
const probe_ms: u32 = 700;
const stop_ms: u32 = 5000;

pub const State = enum { running, stopped, attention };

/// How a new session's background service is started.
///
/// `in_process` forks directly and is what the standalone CLI uses. A live
/// service must not do that: forking a running server would inherit its signal
/// pipe, PTY descriptors and pool state, and would leave the new daemon as its
/// child to reap. `child_process` re-runs the installed binary's own
/// `server start` entry point instead, which forks, detaches and exits, so the
/// new session's service shares nothing with the session that asked for it.
pub const Launcher = enum { in_process, child_process };

/// One named session as `#/$defs/session` describes it. serverID/serverEpoch are
/// present only for a session whose live service actually answered.
pub const Session = struct {
    sessionID: [36]u8,
    /// The name is stored inline so a Session outlives the listing arena that
    /// produced it; registry results are passed across scopes constantly.
    name_bytes: [maximum_name]u8 = undefined,
    name_length: usize = 0,
    state: State,
    serverID: ?[36]u8 = null,
    serverEpoch: ?[36]u8 = null,

    pub fn name(self: *const Session) []const u8 {
        return self.name_bytes[0..self.name_length];
    }

    /// Builds the descriptor for an already-running local service.
    pub fn withKnown(session_id: [36]u8, server_id: [36]u8, server_epoch: [36]u8, value: []const u8) Session {
        return withName(.{ .sessionID = session_id, .state = .running, .serverID = server_id, .serverEpoch = server_epoch }, value);
    }

    fn withName(session: Session, value: []const u8) Session {
        var result = session;
        result.name_length = value.len;
        @memcpy(result.name_bytes[0..value.len], value);
        return result;
    }
};

/// Owns the names and the session array returned by `list`.
pub const Listing = struct {
    arena: std.heap.ArenaAllocator,
    sessions: []Session,
    pub fn deinit(self: *Listing) void {
        self.arena.deinit();
        self.* = undefined;
    }
};

/// The named-session registry is the private state parent directory: one
/// subdirectory per session, each holding that session's own lock, identity,
/// idempotency log and layout snapshot. There is no separate index file, so a
/// crashed writer can never desynchronize the registry from reality.
///
/// A directory without a committed `identity.bin` was never a started session
/// and is deliberately ignored: the protocol requires a stable sessionID, and
/// inventing one for an empty directory would let two listings disagree.
/// Registry view over one private state parent.
///
/// `live` is the identity of the service that is currently serving the request,
/// when the caller is itself one of the registry's sessions. It answers for
/// itself from that value instead of connecting to its own control socket,
/// which it cannot do while it is busy handling this very request.
pub const Registry = struct {
    parent: std.fs.Dir,
    launcher: Launcher = .in_process,
    live: ?Session = null,

    pub fn list(self: Registry, allocator: std.mem.Allocator) !Listing {
        return listIn(allocator, self.parent, self.live);
    }

    pub fn describe(self: Registry, allocator: std.mem.Allocator, name: []const u8) !Session {
        return describeIn(allocator, self.parent, name, self.live);
    }

    pub fn create(self: Registry, allocator: std.mem.Allocator, name: []const u8) !Session {
        return createIn(allocator, self.parent, name, self.launcher);
    }

    pub fn attach(self: Registry, allocator: std.mem.Allocator, session_id: [36]u8) !Session {
        var listing = try self.list(allocator);
        defer listing.deinit();
        return locate(&listing, session_id) orelse error.SessionNotFound;
    }

    /// Stops one session through its own `server.stop` RPC. Never signals a PID
    /// and never touches another session. The layout snapshot is deliberately
    /// left on disk: stopping keeps the structure, only deleting discards it.
    pub fn stop(self: Registry, allocator: std.mem.Allocator, session_id: [36]u8) !Session {
        const found = try self.attach(allocator, session_id);
        // A service cannot drive its own stop handshake from inside a request it
        // is still answering; that is what the session-scope server.stop is for.
        if (self.isSelf(session_id)) return error.SelfSessionTarget;
        if (found.state != .running) return found;
        const reply = try client.stop(allocator, self.parent, found.name(), stop_ms);
        allocator.free(reply.bytes);
        return self.describe(allocator, found.name());
    }

    /// Deletes a stopped session's state directory. A running session is refused
    /// so a delete can never orphan live processes, and only the named session's
    /// own directory is removed.
    pub fn delete(self: Registry, allocator: std.mem.Allocator, session_id: [36]u8) !bool {
        const found = try self.attach(allocator, session_id);
        if (found.state != .stopped) return error.SessionRunning;
        var dir = try StateDirectory.openExisting(self.parent, found.name());
        {
            defer dir.close();
            // Refuse anything that is not this service's own flat state files; an
            // unexpected subdirectory means the caller aimed at the wrong path.
            var iterator = dir.iterate();
            while (try iterator.next()) |entry| {
                if (entry.kind != .file) return error.UnexpectedSessionContent;
            }
            var again = dir.iterate();
            while (try again.next()) |entry| try dir.deleteFile(entry.name);
        }
        try self.parent.deleteDir(found.name());
        return true;
    }

    fn isSelf(self: Registry, session_id: [36]u8) bool {
        const live = self.live orelse return false;
        return std.mem.eql(u8, &live.sessionID, &session_id);
    }
};

fn listIn(allocator: std.mem.Allocator, parent: std.fs.Dir, live: ?Session) !Listing {
    try checkParent(parent);
    var arena = std.heap.ArenaAllocator.init(allocator);
    errdefer arena.deinit();
    const temporary = arena.allocator();
    var found: std.ArrayList(Session) = .empty;
    var iterator = parent.iterate();
    while (try iterator.next()) |entry| {
        if (entry.kind != .directory) continue;
        validateName(entry.name) catch continue;
        if (found.items.len >= maximum_sessions) return error.ResourceLimit;
        var name_bytes: [maximum_name]u8 = undefined;
        @memcpy(name_bytes[0..entry.name.len], entry.name);
        const session = describeIn(allocator, parent, name_bytes[0..entry.name.len], live) catch continue;
        try found.append(temporary, session);
    }
    return .{ .arena = arena, .sessions = try found.toOwnedSlice(temporary) };
}

/// Resolves the real state of one named session. Running is only reported after
/// the service answered on its own socket with a matching sessionID; directory
/// presence alone never counts as running.
fn describeIn(allocator: std.mem.Allocator, parent: std.fs.Dir, name: []const u8, live: ?Session) !Session {
    var dir = try StateDirectory.openExisting(parent, name);
    var identity: ?identities.Identity = null;
    {
        defer dir.close();
        identity = identities.readExisting(dir) catch return error.SessionUnreadable;
    }
    const value = identity orelse return error.SessionNotStarted;
    var session = Session.withName(.{ .sessionID = identities.uuidText(value.session_id), .state = .stopped }, name);
    if (live) |current| {
        if (std.mem.eql(u8, &current.sessionID, &session.sessionID)) {
            session.state = .running;
            session.serverID = current.serverID;
            session.serverEpoch = current.serverEpoch;
            return session;
        }
    }
    const reply = client.status(allocator, parent, name, probe_ms) catch |err| switch (err) {
        // No socket, or a stale one nobody listens on: the service is gone.
        error.FileNotFound, error.ConnectionRefused, error.ServiceDisconnected => return session,
        // Anything else means a service may exist but could not be verified.
        else => {
            session.state = .attention;
            return session;
        },
    };
    defer allocator.free(reply.bytes);
    if (reply.is_error) {
        session.state = .attention;
        return session;
    }
    const parsed = std.json.parseFromSlice(std.json.Value, allocator, reply.bytes, .{}) catch {
        session.state = .attention;
        return session;
    };
    defer parsed.deinit();
    const target = parsed.value.object.get("target") orelse {
        session.state = .attention;
        return session;
    };
    const server_id = text(target, "serverID") orelse return attention(&session);
    const server_epoch = text(target, "serverEpoch") orelse return attention(&session);
    const session_id = text(target, "sessionID") orelse return attention(&session);
    // A live service that does not present the committed sessionID is not this
    // session; report attention rather than adopting its identity.
    if (!std.mem.eql(u8, session_id, &session.sessionID)) return attention(&session);
    session.state = .running;
    session.serverID = server_id[0..36].*;
    session.serverEpoch = server_epoch[0..36].*;
    return session;
}

fn attention(session: *Session) Session {
    session.state = .attention;
    return session.*;
}

fn text(value: std.json.Value, key: []const u8) ?[]const u8 {
    if (value != .object) return null;
    const item = value.object.get(key) orelse return null;
    if (item != .string or item.string.len != 36) return null;
    return item.string;
}

/// Creates a named session and starts its own independent background service.
/// Each session gets its own directory, lock, socket and process, so a fault in
/// one session cannot reach another.
fn createIn(allocator: std.mem.Allocator, parent: std.fs.Dir, name: []const u8, launcher: Launcher) !Session {
    try checkParent(parent);
    try validateName(name);
    if (existing(parent, name)) return error.SessionExists;
    var count: usize = 0;
    var iterator = parent.iterate();
    while (try iterator.next()) |entry| {
        if (entry.kind == .directory) count += 1;
    }
    if (count >= maximum_sessions) return error.ResourceLimit;
    switch (launcher) {
        .in_process => {
            const report = try launch.startAt(allocator, parent, name);
            allocator.free(report);
        },
        .child_process => try startDetached(allocator, parent, name),
    }
    return describeIn(allocator, parent, name, null);
}

/// Runs the installed binary's own background launcher and waits for it. That
/// process performs the fork/detach and exits, so the new daemon is re-parented
/// away and never becomes a zombie child of the caller.
fn startDetached(allocator: std.mem.Allocator, parent: std.fs.Dir, name: []const u8) !void {
    var buffer: [std.fs.max_path_bytes]u8 = undefined;
    const parent_path = try parent.realpath(".", &buffer);
    const executable = try std.fs.selfExePathAlloc(allocator);
    defer allocator.free(executable);
    var child = std.process.Child.init(&.{ executable, "server", "start", parent_path, name }, allocator);
    child.stdin_behavior = .Ignore;
    child.stdout_behavior = .Ignore;
    child.stderr_behavior = .Ignore;
    const term = try child.spawnAndWait();
    if (term != .Exited or term.Exited != 0) return error.StartupOutcomeUnknown;
}

fn existing(parent: std.fs.Dir, name: []const u8) bool {
    _ = std.posix.fstatat(parent.fd, name, std.posix.AT.SYMLINK_NOFOLLOW) catch return false;
    return true;
}

fn locate(listing: *Listing, session_id: [36]u8) ?Session {
    for (listing.sessions) |session| {
        if (std.mem.eql(u8, &session.sessionID, &session_id)) return session;
    }
    return null;
}

fn checkParent(parent: std.fs.Dir) !void {
    const info = try std.posix.fstat(parent.fd);
    if (info.uid != std.posix.geteuid() or info.mode & 0o077 != 0) return error.UnsafeStateParent;
}

/// Registry names are single, printable path components. `.` and `..` would
/// escape the registry, and control characters would make listings ambiguous.
pub fn validateName(name: []const u8) !void {
    if (name.len == 0 or name.len > maximum_name) return error.InvalidSessionName;
    if (std.mem.eql(u8, name, ".") or std.mem.eql(u8, name, "..")) return error.InvalidSessionName;
    for (name) |byte| {
        if (byte == '/' or byte == '\\' or byte < 0x20 or byte == 0x7f) return error.InvalidSessionName;
    }
    StateDirectory.validateName(name) catch return error.InvalidSessionName;
}

test "session registry names reject traversal separators and control bytes" {
    for ([_][]const u8{ "", ".", "..", "a/b", "a\\b", "a\x00b", "a\x1bb" }) |name|
        try std.testing.expectError(error.InvalidSessionName, validateName(name));
    try validateName("work");
    try validateName("a" ** 128);
    try std.testing.expectError(error.InvalidSessionName, validateName("a" ** 129));
}

test "session registry ignores directories that never committed an identity" {
    var tmp = std.testing.tmpDir(.{});
    defer tmp.cleanup();
    try std.posix.fchmod(tmp.dir.fd, 0o700);
    try std.posix.mkdirat(tmp.dir.fd, "unstarted", 0o700);
    const view = Registry{ .parent = tmp.dir };
    var listing = try view.list(std.testing.allocator);
    defer listing.deinit();
    try std.testing.expectEqual(@as(usize, 0), listing.sessions.len);
}

File: SessionRuntime/tests/session_registry.py
"""Real named-session registry integration.

Two named sessions are actually created as independent background services in a
private state parent. Every assertion is about observed reality: real distinct
identities, a real stop that leaves the sibling running, a delete refused while
the session is alive, and the layout snapshot still on disk after a stop.
"""
import json
import os
from pathlib import Path
import shutil
import signal
import socket
import struct
import subprocess
import sys
import tempfile
import time
import uuid

binary = str(Path(sys.argv[1]).resolve())
TIMEOUT = 20.0


def run(*arguments):
    result = subprocess.run([binary, *arguments], capture_output=True, timeout=TIMEOUT)
    assert result.stdout, (arguments, result.stderr[-400:])
    return result.returncode, json.loads(result.stdout)


def packet(value):
    data = json.dumps(value).encode()
    return struct.pack(">BI", 1, len(data)) + data


def receive(peer, deadline):
    def exact(count):
        data = bytearray()
        while len(data) < count:
            remaining = deadline - time.monotonic()
            assert remaining > 0, "frame deadline exceeded"
            peer.settimeout(remaining)
            part = peer.recv(count - len(data))
            assert part, "premature connection close"
            data.extend(part)
        return bytes(data)
    kind, size = struct.unpack(">BI", exact(5))
    assert kind == 1 and 0 < size <= 1024 * 1024, (kind, size)
    return json.loads(exact(size))


def result(response):
    assert "error" not in response, response
    return response["result"]


def failure(code, response):
    assert response.get("error", {}).get("code") == code, response


parent = tempfile.mkdtemp(prefix="aster-registry-", dir="/tmp")
os.chmod(parent, 0o700)
started = []
try:
    code, alpha = run("session", "create", parent, "alpha")
    assert code == 0, alpha
    alpha = result(alpha)
    started.append("alpha")
    code, beta = run("session", "create", parent, "beta")
    assert code == 0, beta
    beta = result(beta)
    started.append("beta")

    # Independent services: distinct persistent identities and distinct
    # incarnations. A shared identity would mean one process served both names.
    assert alpha["sessionID"] != beta["sessionID"], (alpha, beta)
    assert alpha["serverID"] != beta["serverID"], (alpha, beta)
    assert alpha["serverEpoch"] != beta["serverEpoch"], (alpha, beta)
    assert alpha["state"] == beta["state"] == "running"

    code, listing = run("session", "list", parent)
    assert code == 0
    sessions = {item["name"]: item for item in result(listing)["sessions"]}
    assert set(sessions) == {"alpha", "beta"}, sessions
    assert all(item["state"] == "running" for item in sessions.values()), sessions
    assert sessions["alpha"]["sessionID"] == alpha["sessionID"]

    code, attached = run("session", "attach", parent, beta["sessionID"])
    assert code == 0 and result(attached)["serverEpoch"] == beta["serverEpoch"], attached

    # Give each session a real layout so "stop keeps the snapshot" is observable.
    for name in ("alpha", "beta"):
        code, created = run("workspace", "create", parent, name, "--expected-revision", "0",
                            "--title", name, "--cwd", "/tmp", "--", "/bin/sh", "-c", "sleep 300")
        assert code == 0, created
        assert (Path(parent) / name / "layout.json").exists()

    # Deleting a live session must be refused, and must not disturb it.
    code, refused = run("session", "delete", parent, alpha["sessionID"])
    assert code == 1
    failure("session_running", refused)
    code, still = run("session", "attach", parent, alpha["sessionID"])
    assert result(still)["serverEpoch"] == alpha["serverEpoch"], still

    # Registry scope over a session's own control socket: the same registry, and
    # a request with no target, served by whichever session answered.
    endpoint = Path(parent) / "beta/control.sock"
    peer = socket.socket(socket.AF_UNIX)
    try:
        peer.connect(str(endpoint))
        deadline = time.monotonic() + TIMEOUT
        hello = receive(peer, deadline)
        assert hello["type"] == "hello"
        assert {"session_snapshot", "workspace_mutation"} <= set(hello["capabilities"]), hello
        request = dict(type="request", requestID=str(uuid.uuid4()), clientID=str(uuid.uuid4()),
                       scope="registry", operation="session.list", params={})
        peer.sendall(packet(request))
        response = receive(peer, deadline)
        assert response["requestID"] == request["requestID"], response
        assert "target" not in response and "revision" not in response, response
        over_socket = {item["name"] for item in response["result"]["sessions"]}
        assert over_socket == {"alpha", "beta"}, response

        def registry_call(operation, params):
            message = dict(type="request", requestID=str(uuid.uuid4()), clientID=str(uuid.uuid4()),
                           scope="registry", operation=operation, params=params)
            if operation in {"session.create", "session.stop", "session.delete"}:
                message["createdAtUnixMs"] = time.time_ns() // 1_000_000
            peer.sendall(packet(message))
            reply = receive(peer, time.monotonic() + TIMEOUT)
            assert reply["requestID"] == message["requestID"], reply
            return reply

        # A live service can create, stop and delete a sibling session without
        # forking itself; the new service must be a genuinely separate instance.
        spawned = result(registry_call("session.create", {"name": "gamma"}))
        started.append("gamma")
        assert spawned["state"] == "running" and spawned["serverID"] != beta["serverID"], spawned
        assert result(registry_call("session.stop", {"sessionID": spawned["sessionID"]}))["state"] == "stopped"
        started.remove("gamma")
        assert result(registry_call("session.delete", {"sessionID": spawned["sessionID"]}))["deleted"] is True
        assert not (Path(parent) / "gamma").exists()
        # A session cannot stop itself through the registry: that handshake is
        # the session-scope server.stop, not a registry action.
        failure("invalid_request", registry_call("session.stop", {"sessionID": beta["sessionID"]}))
        # Beta served all of that and is untouched.
        assert result(registry_call("session.attach", {"sessionID": beta["sessionID"]}))["serverEpoch"] == beta["serverEpoch"]
    finally:
        peer.close()

    code, stopped = run("session", "stop", parent, alpha["sessionID"])
    assert code == 0, stopped
    assert result(stopped)["state"] == "stopped", stopped
    started.remove("alpha")

    # Stop is scoped to one session: the sibling keeps the same incarnation and
    # the stopped session keeps its layout snapshot on disk.
    code, sibling = run("session", "attach", parent, beta["sessionID"])
    assert result(sibling)["state"] == "running", sibling
    assert result(sibling)["serverEpoch"] == beta["serverEpoch"], sibling
    assert (Path(parent) / "alpha/layout.json").exists(), "stop must preserve the layout snapshot"
    assert not (Path(parent) / "alpha/control.sock").exists()

    code, deleted = run("session", "delete", parent, alpha["sessionID"])
    assert code == 0 and result(deleted)["deleted"] is True, deleted
    assert not (Path(parent) / "alpha").exists()

    code, remaining = run("session", "list", parent)
    names = {item["name"] for item in result(remaining)["sessions"]}
    assert names == {"beta"}, remaining
    code, unknown = run("session", "attach", parent, alpha["sessionID"])
    assert code == 1
    failure("session_not_found", unknown)

    for name in ("", ".", "..", "a/b"):
        code, rejected = run("session", "create", parent, name)
        assert code == 1, rejected
        assert rejected.get("error", {}).get("code") in {"invalid_request"} or rejected.get("code"), rejected
finally:
    for name in started:
        try:
            subprocess.run([binary, "session", "stop", parent, name], capture_output=True, timeout=TIMEOUT)
        except Exception:
            pass
    shutil.rmtree(parent, ignore_errors=True)

print("session registry: independent services; stop kept layout and sibling; running delete refused")
